ledger-service/http-json/src/itlib/scala/http/AbstractHttpServiceIntegrationTest.scala

```diff
@@ -774,12 +772,172 @@ abstract class QueryStoreAndAuthDependentIntegrationTest
+
+      // we assume Disclosed is in the main dalf
+      lazy val inferredPkgId = {
+        import com.daml.lf.{archive, typesig}
+        val dar =
+          archive.UniversalArchiveReader.assertReadFile(AbstractHttpServiceIntegrationTestFuns.dar2)
+        typesig.PackageSignature.read(dar.main)._2.packageId
+      }
+
+      def inDar2Main(
+          tid: domain.ContractTypeId.Template.OptionalPkg
+      ): domain.ContractTypeId.Template.RequiredPkg =
+        tid.copy(packageId = inferredPkgId)
+
+      lazy val ToDisclose = inDar2Main(TpId.Disclosure.ToDisclose)
+
+      lazy val (_, toDiscloseVA) =
+        VA.record(lfIdentifier(ToDisclose), ShRecord(owner = VAx.partyDomain, junk = VA.text))
+
+      val (_, viewportVA) =
+        VA.record(
+          lfIdentifier(TpId.Disclosure.Viewport.copy(packageId = "ignored")),
+          ShRecord(owner = VAx.partyDomain),
+        )
+
+      val (_, checkVisibilityVA) =
+        VA.record(
+          Ref.Identifier assertFromString "ignored:Disclosure:CheckVisibility",
+          ShRecord(disclosed = VAx.contractIdDomain),
+        )
+
+      final case class ContractToDisclose(
+          alice: domain.Party,
+          toDiscloseCid: domain.ContractId,
+          toDisclosePayload: lav1.value.Record,
+          ctMetadata: DC.Metadata,
+      )
+
+      def contractToDisclose(
+          fixture: HttpServiceTestFixtureData,
+          junkMessage: String,
+      ): Future[ContractToDisclose] = for {
+        (alice, jwt, applicationId, _) <- fixture.getUniquePartyTokenAppIdAndAuthHeaders("Alice")
+        // we're using the ledger API for the initial create because timestamp
+        // is required in the metadata
+        toDisclosePayload = argToApi(toDiscloseVA)(ShRecord(owner = alice, junk = junkMessage))
+        createCommand = util.Commands.create(
+          refApiIdentifier(ToDisclose),
+          toDisclosePayload,
+        )
+        initialCreate = SubmitAndWaitRequest(
+          Some(
+            Commands(
+              commandId = uniqueCommandId().unwrap,
+              applicationId = applicationId.unwrap,
+              actAs = domain.Party unsubst Seq(alice),
+              commands = Seq(Command(createCommand)),
+            )
+          )
+        )
+        createResp <- fixture.client.commandServiceClient
+          .submitAndWaitForTransaction(initialCreate, Some(jwt.value))
+        (toDiscloseCid, ctMetadata) = inside(createResp.transaction) { case Some(tx) =>
+          inside(tx.events) { case Seq(lav1.event.Event(lav1.event.Event.Event.Created(ce))) =>
+            (
+              domain.ContractId(ce.contractId),
+              inside(ce.metadata map DC.Metadata.fromLedgerApi) { case Some(\/-(metadata)) =>
+                metadata
+              },
+            )
+          }
+        }
+      } yield ContractToDisclose(alice, toDiscloseCid, toDisclosePayload, ctMetadata)
+
+      "using decoded payload" in withHttpService { fixture =>
```

Comment:
afaict you're missing tests for:
1. Passing a blob. This should be tested using an interface.
2. Passing a contract with a key.

Reply:
No I don't mean exercising an interface choice at the top-level. I mean passing an interface contract in disclosed contracts. That's the case blobs are intended for because decoding the payload introduces coupling to at least the lf version. The top-level choice you're exercising can still be on a template. The template id in the disclosed contract points to the template not the interface in that case.

Reply:
We can sufficiently exercise this case just by exercising the `payloadBlob` case, because we don't decode `payloadBlob`s to their records; interfaces don't enter into it.

Reply:
#16624